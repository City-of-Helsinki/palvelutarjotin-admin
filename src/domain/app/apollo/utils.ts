import apolloClient from './apolloClient';

export const clearApolloCache = async () => {
  apolloClient.stop();
  await Promise.all([apolloClient.resetStore(), apolloClient.clearStore()]);
};
