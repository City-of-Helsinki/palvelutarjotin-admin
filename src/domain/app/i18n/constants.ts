export enum VALIDATION_MESSAGE_KEYS {
  EMAIL = 'form.validation.string.email',
  NUMBER_MAX = 'form.validation.number.max',
  NUMBER_MIN = 'form.validation.number.min',
  NUMBER_REQUIRED = 'form.validation.number.required',
  STRING_MAX = 'form.validation.string.max',
  STRING_MIN = 'form.validation.string.min',
  STRING_REQUIRED = 'form.validation.string.required',
  URL = 'form.validation.string.url',
}
