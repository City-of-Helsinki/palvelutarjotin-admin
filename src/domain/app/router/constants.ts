import type { BrowserRouterProps, MemoryRouterProps } from 'react-router-dom';

export const DEFAULT_ROUTER_PROPS = {
  future: {
    v7_startTransition: true,
  },
} as const satisfies BrowserRouterProps satisfies MemoryRouterProps;
