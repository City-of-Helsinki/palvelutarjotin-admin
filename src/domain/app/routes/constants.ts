export enum ROUTES {
  CALLBACK = '/callback',
  CREATE_EVENT = '/events/create',
  CREATE_OCCURRENCE = '/events/:id/occurrences/create',
  EDIT_OCCURRENCE = '/events/:id/occurrences/:occurrenceId/edit',
  EDIT_EVENT = '/events/:id/edit',
  ENROLLMENTS = '/events/:id/occurrences/:occurrenceId/enrollments',
  EVENT_DETAILS = '/events/:id',
  HOME = '/',
  OCCURRENCES = '/events/:id/occurrences',
  SILENT_CALLBACK = '/silent-callback',
}
