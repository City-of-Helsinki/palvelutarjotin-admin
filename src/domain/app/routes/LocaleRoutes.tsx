import React from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { Route, RouteComponentProps, Switch } from 'react-router';

import { SUPPORT_LANGUAGES } from '../../../constants';
import { tokenFetched } from '../../auth/actions';
import { getApiToken } from '../../auth/authenticate';
import { apiTokenSelector, userSelector } from '../../auth/selectors';
import CreateEventPage from '../../event/CreateEventPage';
import EventDetailsPage from '../../event/EventDetailsPage';
import EventsPage from '../../events/EventsPage';
import NotFoundPage from '../../notFound/NotFoundPage';
import PageLayout from '../layout/PageLayout';
import { ROUTES } from './constants';

const LocaleRoutes: React.FC<RouteComponentProps<{
  locale: SUPPORT_LANGUAGES;
}>> = ({
  match: {
    params: { locale },
  },
}) => {
  const { i18n } = useTranslation();
  const apiToken = useSelector(apiTokenSelector);
  const dispatch = useDispatch();
  const user = useSelector(userSelector);

  React.useEffect(() => {
    if (apiToken) {
      // Skip token fetch if token already existed
      dispatch(tokenFetched());

      // If no token but access token is ready for exchange
      // start to fetch apiToken
    } else if (user?.access_token) {
      dispatch(getApiToken(user.access_token));
    }
  }, [apiToken, dispatch, user]);

  React.useEffect(() => {
    i18n.changeLanguage(locale);
  }, [i18n, locale]);

  return (
    <PageLayout>
      <Switch>
        <Route exact path={`/${locale}`} component={EventsPage} />
        <Route
          exact
          path={`/${locale}${ROUTES.CREATE_EVENT}`}
          component={CreateEventPage}
        />
        <Route
          exact
          path={`/${locale}${ROUTES.EVENT_DETAILS}`}
          component={EventDetailsPage}
        />
        <Route component={NotFoundPage} />
      </Switch>
    </PageLayout>
  );
};

export default LocaleRoutes;
