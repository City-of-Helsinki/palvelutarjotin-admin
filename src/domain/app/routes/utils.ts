import { ROUTES } from './constants';

export const getCmsPath = (slug?: string | null) => {
  return ROUTES.CMS_PAGE.replace('/:slug', slug ?? '');
};
