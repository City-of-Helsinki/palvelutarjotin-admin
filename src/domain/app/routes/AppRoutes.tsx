import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { Navigate, Route, Routes, useLocation } from 'react-router-dom';

import { ROUTES } from './constants';
import LocaleRoutes from './LocaleRoutes';
import SilentRenewRedirect from './SilentRenewRedirect';
import KultusAdminHDSLoginCallbackHandler from '../../auth/KultusAdminHDSLoginCallbackHandler';

const NavigateToLocalePath = () => {
  const {
    i18n: { language: currentLocale },
  } = useTranslation();
  const location = useLocation();
  return (
    <Navigate
      to={`/${currentLocale}${location.pathname}${location.search}`}
      replace
    />
  );
};

const AppRoutes = () => {
  const { i18n } = useTranslation();
  const currentLocale = i18n.language;

  const fiLocaleRoutes = <LocaleRoutes locale="fi" />;
  const svLocaleRoutes = <LocaleRoutes locale="sv" />;
  const enLocaleRoutes = <LocaleRoutes locale="en" />;

  return (
    <Routes>
      <Route path="/" element={<Navigate to={`/${currentLocale}`} replace />} />
      <Route
        path={ROUTES.CALLBACK}
        element={<KultusAdminHDSLoginCallbackHandler />}
      />
      <Route path={ROUTES.SILENT_CALLBACK} element={<SilentRenewRedirect />} />
      <Route path={ROUTES.SILENT_CALLBACK_HTML} />
      <Route path="/fi">
        <Route index element={fiLocaleRoutes} />
        <Route path="*" element={fiLocaleRoutes} />
      </Route>
      <Route path="/sv">
        <Route index element={svLocaleRoutes} />
        <Route path="*" element={svLocaleRoutes} />
      </Route>
      <Route path="/en">
        <Route index element={enLocaleRoutes} />
        <Route path="*" element={enLocaleRoutes} />
      </Route>
      <Route path="*" element={<NavigateToLocalePath />} />
    </Routes>
  );
};

export default AppRoutes;
