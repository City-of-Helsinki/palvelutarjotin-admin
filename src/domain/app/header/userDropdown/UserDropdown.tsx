import { IconAngleRight, IconArrowRight, IconUser } from 'hds-react';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import MenuDropdown from '../../../../common/components/menuDropdown/MenuDropdown';
import {
  MyProfileQuery,
  OrganisationNodeFieldsFragment,
} from '../../../../generated/graphql';
import useHistory from '../../../../hooks/useHistory';
import { logoutTunnistamo } from '../../../auth/authenticate';
import { setActiveOrganisation } from '../../../organisation/actions';
import { activeOrganisationSelector } from '../../../organisation/selector';
import { ROUTES } from '../../routes/constants';
import styles from './userDropdown.module.scss';

const MENU_ITEM_VALUES = {
  LOGOUT: 'logout',
  OPEN_MY_PROFILE: 'open_my_profile',
};

interface Props {
  myProfileData: MyProfileQuery;
}

const UserDropdown: React.FC<Props> = ({ myProfileData }) => {
  const { t } = useTranslation();
  const activeOrganisation = useSelector(activeOrganisationSelector);
  const dispatch = useDispatch();
  const history = useHistory();

  const organisations: OrganisationNodeFieldsFragment[] =
    myProfileData.myProfile?.organisations.edges.map((edge) => ({
      ...(edge?.node as OrganisationNodeFieldsFragment),
    })) || [];

  const goToEditMyProfile = () => {
    history.pushWithLocale(ROUTES.MY_PROFILE);
  };

  const logout = () => {
    logoutTunnistamo();
  };

  const changeActiveOrganisation = (id: string) => {
    dispatch(setActiveOrganisation(id));
  };

  return (
    <MenuDropdown
      buttonAriaLabel={t('header.userMenu.ariaLabelButton')}
      buttonText={myProfileData.myProfile?.name || ''}
      icon={<IconUser />}
      items={[
        {
          icon: <IconUser />,
          onClick: goToEditMyProfile,
          text: t('header.userMenu.openMyProfile'),
          value: MENU_ITEM_VALUES.OPEN_MY_PROFILE,
        },
        ...organisations?.map((organisation) => ({
          className:
            activeOrganisation === organisation.id
              ? styles.activeOrganisation
              : undefined,
          icon: <IconAngleRight />,
          onClick: changeActiveOrganisation,
          text: organisation.name || '',
          value: organisation.id || '',
        })),
        {
          icon: <IconArrowRight />,
          onClick: logout,
          text: t('header.userMenu.logout'),
          value: MENU_ITEM_VALUES.LOGOUT,
        },
      ]}
    />
  );
};

export default UserDropdown;
