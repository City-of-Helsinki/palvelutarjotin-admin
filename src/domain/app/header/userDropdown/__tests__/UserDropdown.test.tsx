import * as React from 'react';
import * as reduxFuncs from 'react-redux';

import {
  fakeOrganisations,
  fakePerson,
} from '../../../../../utils/mockDataUtils';
import { render, screen, userEvent } from '../../../../../utils/testUtils';
import { ORGANISATION_ACTIONS } from '../../../../organisation/constants';
import { ROUTES } from '../../../routes/constants';
import UserDropdown from '../UserDropdown';

const personName = 'Testi Testaaja';
const organisationId = 'organisationId1';

const profileMock = fakePerson({
  organisations: fakeOrganisations(3, [
    { name: 'Organisaatio 1' },
    { name: 'Organisaatio 2' },
    { name: 'Organisaatio 3', id: organisationId },
  ]),
  name: personName,
});

it('matches snapshot', () => {
  const { container } = render(
    <UserDropdown myProfileData={{ myProfile: profileMock }} />
  );

  userEvent.click(
    screen.getByRole('button', { name: 'Avaa käyttäjän valikko' })
  );

  expect(container).toMatchSnapshot();
});

it('renders dropdown correctly', () => {
  const dispatchMock = jest.fn();
  jest.spyOn(reduxFuncs, 'useDispatch').mockReturnValue(dispatchMock);

  const { history } = render(
    <UserDropdown myProfileData={{ myProfile: profileMock }} />
  );

  const push = jest.spyOn(history, 'push');

  userEvent.click(
    screen.getByRole('button', { name: 'Avaa käyttäjän valikko' })
  );

  expect(
    screen.getByRole('menuitem', {
      name: 'Organisaatio 1',
    })
  ).toBeInTheDocument();
  expect(
    screen.getByRole('menuitem', {
      name: 'Organisaatio 2',
    })
  ).toBeInTheDocument();
  expect(
    screen.getByRole('menuitem', {
      name: 'Organisaatio 3',
    })
  ).toBeInTheDocument();

  expect(
    screen.queryByRole('menuitem', {
      name: 'Kirjaudu ulos',
    })
  ).toBeInTheDocument();

  userEvent.click(
    screen.getByRole('menuitem', {
      name: 'Organisaatio 3',
    })
  );

  expect(dispatchMock).toHaveBeenCalledWith({
    payload: organisationId,
    type: ORGANISATION_ACTIONS.SET_ACTIVE_ORGANISATION,
  });

  userEvent.click(
    screen.getByRole('menuitem', {
      name: 'Omat tiedot',
    })
  );

  expect(push).toHaveBeenCalledWith(`/fi${ROUTES.MY_PROFILE}`);
});
