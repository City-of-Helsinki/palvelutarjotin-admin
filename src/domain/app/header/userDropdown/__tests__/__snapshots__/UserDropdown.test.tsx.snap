// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="menuDropdown"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="Avaa käyttäjän valikko"
      class="dropdownButton"
      type="button"
    >
      <div
        class="iconWrapper"
      >
        <svg
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M16.267 13a4.733 4.733 0 014.728 4.517l.005.216V22H3v-4.267a4.733 4.733 0 014.517-4.728L7.733 13h8.534zm-8.534 2a2.733 2.733 0 00-2.728 2.567L5 17.733V20h14v-2.267c0-1.416-1.107-2.636-2.46-2.726l-.163-.005L7.733 15zM12 2a5 5 0 110 10 5 5 0 010-10zm0 2a3 3 0 100 6 3 3 0 000-6z"
              fill="currentColor"
            />
          </g>
        </svg>
      </div>
      <div
        class="textWrapper"
      >
        Testi Testaaja
        <svg
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik iconAngle iconAngleUp"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </div>
    </button>
    <ul
      class="dropdownMenu isOpen"
      role="menu"
    >
      <li
        class="menuItem"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="menuItemIconWrapper"
        >
          <svg
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M16.267 13a4.733 4.733 0 014.728 4.517l.005.216V22H3v-4.267a4.733 4.733 0 014.517-4.728L7.733 13h8.534zm-8.534 2a2.733 2.733 0 00-2.728 2.567L5 17.733V20h14v-2.267c0-1.416-1.107-2.636-2.46-2.726l-.163-.005L7.733 15zM12 2a5 5 0 110 10 5 5 0 010-10zm0 2a3 3 0 100 6 3 3 0 000-6z"
                fill="currentColor"
              />
            </g>
          </svg>
        </div>
        Omat tiedot
      </li>
      <li
        class="menuItem"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="menuItemIconWrapper"
        >
          <svg
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <polygon
                points="0 24 0 0 24 0 24 24"
              />
              <polygon
                fill="currentColor"
                points="13.5 12 8.5 7 10 5.5 16.5 12 10 18.5 8.5 17"
              />
            </g>
          </svg>
        </div>
        Organisaatio 1
      </li>
      <li
        class="menuItem"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="menuItemIconWrapper"
        >
          <svg
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <polygon
                points="0 24 0 0 24 0 24 24"
              />
              <polygon
                fill="currentColor"
                points="13.5 12 8.5 7 10 5.5 16.5 12 10 18.5 8.5 17"
              />
            </g>
          </svg>
        </div>
        Organisaatio 2
      </li>
      <li
        class="menuItem"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="menuItemIconWrapper"
        >
          <svg
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <polygon
                points="0 24 0 0 24 0 24 24"
              />
              <polygon
                fill="currentColor"
                points="13.5 12 8.5 7 10 5.5 16.5 12 10 18.5 8.5 17"
              />
            </g>
          </svg>
        </div>
        Organisaatio 3
      </li>
      <li
        class="menuItem"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="menuItemIconWrapper"
        >
          <svg
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <rect
                height="24"
                width="24"
              />
              <polygon
                fill="currentColor"
                points="10.5 5.5 12 7 8 11 20.5 11 20.5 13 8 13 12 17 10.5 18.5 4 12"
                transform="matrix(-1 0 0 1 24.5 0)"
              />
            </g>
          </svg>
        </div>
        Kirjaudu ulos
      </li>
    </ul>
  </div>
</div>
`;
